shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,323 @@
+    }
+
+    fn decode_response(&self, src: &mut BytesMut) -> Result<Option<(usize, HttpHead, usize)>> {
+        let mut headers = [httparse::EMPTY_HEADER; 16];
+        let mut response = httparse::Response::new(&mut headers);
+
+        let amt = match response.parse(&src[0..src.len()]).unwrap() {
+            httparse::Status::Complete(amt) => amt,
+            httparse::Status::Partial => return Ok(None),
+        };
+        match response.version.unwrap() {
+            1 => (),
+            _version => panic!("error!"), // TODO version error
+        }
+
+        let mut builder = Response::builder();
+
+        builder = builder.status(response.code.unwrap());
+
+        let builder_headers = builder.headers_mut().unwrap();
+
+        let response_headers: Vec<(String, Vec<u8>)> = response
+            .headers
+            .iter()
+            .take_while(|h| !h.name.is_empty() && !h.value.is_empty())
+            .map(|h| (h.name.to_owned(), h.value.to_owned()))
+            .collect();
+
+        for (name, value) in response_headers {
+            builder_headers.insert(
+                HeaderName::from_bytes(name.as_bytes()).unwrap(),
+                HeaderValue::from_bytes(&value).unwrap(),
+            );
+        }
+
+        let r = builder.body(()).unwrap();
+        let cl = match r.headers().get(header::CONTENT_LENGTH) {
+            Some(cl) => match atoi::atoi(cl.as_bytes()) {
+                Some(cl) => cl,
+                None => panic!(), // TODO content length error
+            },
+            None => 0, // TODO content length error
+        };
+        let (parts, _) = r.into_parts();
+        Ok(Some((amt, HttpHead::Response(parts), cl)))
+    }
+}
+
+#[derive(Debug)]
+enum State {
+    ParsingResponse,
+    ReadingBody(HttpHead, usize),
+}
+
+impl Decoder for OpenSearchDecoder {
+    type Item = Messages;
+    type Error = CodecReadError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, CodecReadError> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+
+        loop {
+            match std::mem::replace(&mut self.state, State::ParsingResponse) {
+                State::ParsingResponse => {
+                    let decode_res = if self.direction == Direction::Source {
+                        self.decode_request(src).unwrap() // TODO
+                    } else {
+                        self.decode_response(src).unwrap() // TODO
+                    };
+
+                    if let Some((amt, parts, cl)) = decode_res {
+                        self.state = State::ReadingBody(parts, cl);
+                        src.advance(amt);
+                    } else {
+                        return Ok(None);
+                    };
+                }
+                State::ReadingBody(parts, cl) => {
+                    if src.is_empty() {
+                        // sometimes messages will have a content-length but no
```

Comment:
is this behavior you've encountered or something that you are guessing might happen?

Reply:
Behaviour I encountered with OpenSearch

Reply:
This sounds like a bug to me, we should remove this hack and investigate whats going on.

Reply:
I found that the issue occurs on:
```rust
        client
            .indices()
            .exists(IndicesExistsParts::Index(&["test-index"]))
            .send()
            .await,
```

And that if I do:
```rust
    assert_ok_and_get_json(
        client
            .indices()
            .exists(IndicesExistsParts::Index(&["test-index"]))
            .send()
            .await,
    )
    .await;
```

Then I get a panic in `assert_ok_and_get_json`:
```
thread 'opensearch_int_tests::passthrough_standard' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: Http(reqwest::Error { kind: Decode, source: Error("EOF while parsing a value", line: 1, column: 0) }) }', shotover-proxy/tests/opensearch_int_tests/mod.rs:22:47
```

Both with and without shotover.
So seems to be a bug in either opensearch or the rust opensearch driver.

Reply:
Ah! figured it out.
The client sends an HTTP HEAD request
And as per https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD a head request, requests for just the headers without the body.
So we should change this to check for HEAD instead of `src.is_empty()`

Reply:
That would require some shared state between the encode and decode sides of the codecs, as the response to a HEAD request has nothing indicating it is that, it just responds the same as a GET request, but only the headers.

Reply:
sounds like we'll need to add that shared state then.
checking `src.is_empty()` is no good as if we happen to perform the check just before receiving the bytes of the body we will ignore the body entirely.

Reply:
Hmm, I had imagined that case would be very unlikely. Anyway the state wasn't difficult to implement just thought that it was better avoided if possible.